docs/pflask.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PFLASK" "1" "February 2015" "" ""
.
.SH "NAME"
\fBpflask\fR \- the process in the flask
.
.SH "SYNOPSIS"
\fBpflask [OPTIONS] [COMMAND [ARGS\.\.\.]]\fR
.
.SH "DESCRIPTION"
\fBpflask\fR is a simple tool for creating Linux namespace containers\. It can be used for running a command or even booting an OS inside an isolated container, created with the help of Linux namespaces\. It is similar in functionality to \fBchroot(8)\fR, although pflask provides better isolation thanks to the use of namespaces\.
.
.P
Additionally, while most other containerization solutions (LXC, systemd\-nspawn, \.\.\.) are mostly targeted at containing whole systems, pflask can also be used to contain single programs, without the need to create ad\-hoc chroots\.
.
.SH "OPTIONS"
\fB\-m, \-\-mount=<type>,<opts>\fR
.
.P
\~\~\~\~\~\~ Create a new \fI"type"\fR mount point inside the container\. See \fIMOUNT\fR\.
.
.P
\fB\-n, \-\-netif[=<opts>]\fR
.
.P
\~\~\~\~\~\~ Create a new network namespace and, optionally create/move a network interface inside the container\. See \fINETIF\fR\.
.
.P
\fB\-u, \-\-user=<user>\fR
.
.P
\~\~\~\~\~\~ Run the command as \fI"user"\fR inside the container\. If \fB`\-\-no\-userns\fR is not used, this will also create a new user namespace\.
.
.P
\fB\-r, \-\-chroot=<dir>\fR
.
.P
\~\~\~\~\~\~ Use \fI"dir"\fR as root directory inside the container\.
.
.P
\fB\-c, \-\-chdir=<dir>\fR
.
.P
\~\~\~\~\~\~ Change to \fI"dir"\fR inside the container\.
.
.P
\fB\-g, \-\-cgroup=<controller>[,<controller> \.\.\.]\fR
.
.P
\~\~\~\~\~\~ Create new cgroups in the given controllers and move the container inside them\.
.
.P
\fB\-d, \-\-detach\fR
.
.P
\~\~\~\~\~\~ Detach from terminal\.
.
.P
\fB\-a, \-\-attach=<pid>\fR
.
.P
\~\~\~\~\~\~ Attach to the \fI"pid"\fR detached process\. Only a process with the same UID of the detached process can attach to it\. To detach again press \fB^@\fR (Ctrl + @)\.
.
.P
\fB\-s, \-\-setenv=<name>=<value>[,<name>=<value> \.\.\.]\fR
.
.P
\~\~\~\~\~\~ Set additional environment variables\. It takes a comma\-separated list of variables of the form \fBname=value\fR\. This option may be used more than once\.
.
.P
\fB\-k, \-\-keepenv\fR
.
.P
\~\~\~\~\~\~ Do not clear environment (only relevant when used with \fB\-\-chroot\fR)\.
.
.P
\fB\-U, \-\-no\-userns\fR
.
.P
\~\~\~\~\~\~ Disable user namespace\.
.
.P
\fB\-M, \-\-no\-mountns\fR
.
.P
\~\~\~\~\~\~ Disable mount namespace\.
.
.P
\fB\-N, \-\-no\-netns\fR
.
.P
\~\~\~\~\~\~ Disable net namespace\.
.
.P
\fB\-I, \-\-no\-ipcns\fR
.
.P
\~\~\~\~\~\~ Disable IPC namespace\.
.
.P
\fB\-H, \-\-no\-utsns\fR
.
.P
\~\~\~\~\~\~ Disable UTS namespace\.
.
.P
\fB\-P, \-\-no\-pidns\fR
.
.P
\~\~\~\~\~\~ Disable PID namespace\.
.
.SH "MOUNT"
pflask support the following mount point types:
.
.SS "bind"
It bind mounts a directory/file to another directory/file
.
.P
Example: \fB\-\-mount=bind,/source/path,/dest/path\fR
.
.SS "bind\-ro"
Same as \fBbind\fR, but make the mount point read\-only\.
.
.P
Example: \fB\-\-mount=bind\-ro,/source/path,/dest/path\fR
.
.SS "aufs"
It stacks a directory on top of another directory using AuFS\. Note that AuFS doesn\'t support user namespaces, so \fB\-\-no\-userns\fR will also be needed\.
.
.P
Example: \fB\-\-mount=aufs,/overlay/path,/dest/path\fR
.
.SS "overlay"
Like aufs, but uses OverlayFS instead of AuFS\. It either requires Linux 3\.18 or the OverlayFS module installed\. Note that OverlayFS doesn\'t support user namespaces, so \fB\-\-no\-userns\fR will also be needed\.
.
.P
Example: \fB\-\-mount=overlay,/overlay/path,/dest/path\fR
.
.SS "tmp"
It mounts a tmpfs on a directory\.
.
.P
Example: \fB\-\-mount=tmp,/dest/path\fR
.
.SH "NETIF"
When the \fB\-\-netif\fR option is used, pflask will create a new network namespace inside the container\. If the argument is set, the following actions will be taken:
.
.SS "move and rename"
\~\~\~\~\~\~ \fB\-\-netif=<dev>,<name>\fR
.
.P
If the \fI"dev"\fR option is an existing network interface, it will be moved inside the container "as is" and renamed to \fI"name"\fR\. No additional configuration will be applied to it\.
.
.P
Example: \fB\-\-netif=vxlan0,eth0\fR
.
.SS "macvlan"
\~\~\~\~\~\~ \fB\-\-netif=macvlan,<master>,<name>\fR
.
.P
If the \fI"macvlan"\fR option is used, a new network interface of type \fBmacvlan\fR will be created using \fI"master"\fR as master interface, moved inside the container and renamed to \fI"name"\fR\. No additional configuration will be applied to it\.
.
.P
Example: \fB\-\-netif=macvlan,eth0,eth0\fR
.
.SS "veth"
\~\~\~\~\~\~ \fB\-\-netif=veth,<name_outside>,<name_inside>\fR
.
.P
If the \fI"veth"\fR option is used, a new pair of network interfaces of type \fBveth\fR will be created and one of the two moved inside the container\. The twin outside the container will be named \fI"name_outside"\fR, while the twin inside the container will be named \fI"name_inside"\fR\. No additional configuration will be applied to them\.
.
.P
Example: \fB\-\-netif=veth,veth0,eth0\fR
.
.SH "AUTHOR"
Alessandro Ghedini \fIalessandro@ghedini\.me\fR
.
.SH "COPYRIGHT"
Copyright (C) 2013 Alessandro Ghedini \fIalessandro@ghedini\.me\fR
.
.P
This program is released under the 2 clause BSD license\.


src/pflask.c
/*
 * The process in the flask.
 *
 * Copyright (c) 2013, Alessandro Ghedini
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>

#include <sys/syscall.h>

#include <sched.h>
#include <linux/sched.h>

#include <signal.h>
#include <getopt.h>

#include <sys/prctl.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include "dev.h"
#include "pty.h"
#include "user.h"
#include "mount.h"
#include "cgroup.h"
#include "netif.h"
#include "printf.h"
#include "util.h"

static int clone_flags = SIGCHLD      |
                         CLONE_NEWNS  |
                         CLONE_NEWIPC |
                         CLONE_NEWPID |
                         CLONE_NEWUTS;

static const char *short_opts = "+m:n::u:r:c:g:da:s:kUMNIHPh?";

static struct option long_opts[] = {
	{ "mount",     required_argument, NULL, 'm' },
	{ "netif",     optional_argument, NULL, 'n' },
	{ "user",      required_argument, NULL, 'u' },
	{ "chroot",    required_argument, NULL, 'r' },
	{ "chdir",     required_argument, NULL, 'c' },
	{ "cgroup",    required_argument, NULL, 'g' },
	{ "detach",    no_argument,       NULL, 'd' },
	{ "attach",    required_argument, NULL, 'a' },
	{ "setenv",    required_argument, NULL, 's' },
	{ "keepenv",   no_argument,       NULL, 'k' },
	{ "no-userns", no_argument,       NULL, 'U' },
	{ "no-mountns", no_argument,      NULL, 'M' },
	{ "no-netns",  no_argument,       NULL, 'N' },
	{ "no-ipcns",  no_argument,       NULL, 'I' },
	{ "no-utsns",  no_argument,       NULL, 'H' },
	{ "no-pidns",  no_argument,       NULL, 'P' },
	{ "help",      no_argument,       NULL, 'h' },
	{ 0, 0, 0, 0 }
};

static void do_daemonize(void);
static void do_chroot(char *dest);
static pid_t do_clone(void);

static inline void help(void);

int main(int argc, char *argv[]) {
	int rc, i;

	pid_t pid  = -1;
	pid_t ppid = getpid();

	uid_t uid = -1;
	gid_t gid = -1;

	_free_ char *user   = NULL;
	_free_ char *dest   = NULL;
	_free_ char *change = NULL;
	_free_ char *env    = NULL;
	_free_ char *cgroup = NULL;

	_close_ int master_fd = -1;

	char *master_name;

	int detach  = 0;
	int keepenv = 0;

	siginfo_t status;

	while ((rc = getopt_long(argc, argv, short_opts, long_opts, &i)) !=-1) {
		switch (rc) {
			case 'm':
				validate_optlist("--mount", optarg);

				add_mount_from_spec(optarg);
				break;

			case 'n':
				clone_flags |= CLONE_NEWNET;

				if (optarg != NULL) {
					validate_optlist("--netif", optarg);

					add_netif_from_spec(optarg);
				}
				break;

			case 'u':
				clone_flags |= CLONE_NEWUSER;

				freep(&user);

				user = strdup(optarg);
				break;

			case 'r':
				freep(&dest);

				dest = realpath(optarg, NULL);
				if (dest == NULL) sysf_printf("realpath()");
				break;

			case 'c':
				freep(&change);

				change = strdup(optarg);
				break;

			case 'g':
				validate_optlist("--cgroup", optarg);
				validate_cgroup_spec(optarg);

				freep(&change);

				cgroup = strdup(optarg);
				break;

			case 'd':
				detach = 1;
				break;

			case 'a': {
				char *end = NULL;
				pid = strtol(optarg, &end, 10);
				if (*end != '\0')
					fail_printf("Invalid value '%s' for --attach", optarg);
				break;
			}

			case 's': {
				validate_optlist("--setenv", optarg);

				if (env != NULL) {
					char *tmp = env;

					rc = asprintf(&env, "%s,%s", env, optarg);
					if (rc < 0) fail_printf("OOM");

					freep(&tmp);
				} else {
					env = strdup(optarg);
				}

				break;
			}

			case 'k':
				keepenv = 1;
				break;

			case 'U':
				clone_flags &= ~(CLONE_NEWUSER);
				break;

			case 'M':
				clone_flags &= ~(CLONE_NEWNS);
				break;

			case 'N':
				clone_flags &= ~(CLONE_NEWNET);
				break;

			case 'I':
				clone_flags &= ~(CLONE_NEWIPC);
				break;

			case 'H':
				clone_flags &= ~(CLONE_NEWUTS);
				break;

			case 'P':
				clone_flags &= ~(CLONE_NEWPID);
				break;

			case '?':
			case 'h':
				help();
				return 0;
		}
	}

	if (pid != -1) {
		master_fd = recv_pty(pid);
		if (master_fd < 0) fail_printf("Invalid PID '%u'", pid);

		pid = -1;
		goto process_fd;
	}

	if (user == NULL) {
		user = strdup("root");
		if (user == NULL) fail_printf("OOM");
	}

	open_master_pty(&master_fd, &master_name);

	uid = getuid();
	gid = getgid();

	if (detach == 1)
		do_daemonize();

	pid = do_clone();

	if (pid == 0) {
		closep(&master_fd);

		open_slave_pty(master_name);

		rc = setsid();
		if (rc < 0) sysf_printf("setsid()");

		rc = prctl(PR_SET_PDEATHSIG, SIGKILL);
		if (rc < 0) sysf_printf("prctl(PR_SET_PDEATHSIG)");

		if (clone_flags & CLONE_NEWUSER)
			map_user_to_user(uid, gid, user);

		do_cgroup(cgroup, ppid);

		do_mount(dest);

		if (dest != NULL) {
			copy_nodes(dest);

			make_ptmx(dest);

			make_symlinks(dest);

			make_console(dest, master_name);

			do_chroot(dest);
		}

		if (clone_flags & CLONE_NEWNET)
			setup_loopback();

		umask(0022);

		/* TODO: drop capabilities */

		do_user(user);

		if (change != NULL) {
			rc = chdir(change);
			if (rc < 0) sysf_printf("chdir()");
		}

		if (dest != NULL) {
			char *term = getenv("TERM");

			if (keepenv == 0)
				clearenv();

			setenv("PATH", "/usr/sbin:/usr/bin:/sbin:/bin", 1);
			setenv("USER", user, 1);
			setenv("LOGNAME", user, 1);
			setenv("TERM", term, 1);
		}

		if (env != NULL) {
			size_t i, c;

			_free_ char **vars = NULL;

			_free_ char *tmp = strdup(env);
			if (tmp == NULL) fail_printf("OOM");

			c = split_str(tmp, &vars, ",");

			for (i = 0; i < c; i++) {
				rc = putenv(strdup(vars[i]));
				if (rc != 0) sysf_printf("putenv()");
			}
		}

		setenv("container", "pflask", 1);

		if (argc > optind)
			rc = execvpe(argv[optind], argv + optind, environ);
		else
			rc = execle("/bin/bash", "-bash", NULL, environ);

		if (rc < 0) sysf_printf("exec()");
	}

	do_netif(pid);

process_fd:
	if (detach == 1)
		serve_pty(master_fd);
	else
		process_pty(master_fd);

	if (pid == -1)
		return 0;

	kill(pid, SIGKILL);

	rc = waitid(P_PID, pid, &status, WEXITED);
	if (rc < 0) sysf_printf("waitid()");

	switch (status.si_code) {
		case CLD_EXITED:
			if (status.si_status != 0)
				err_printf("Child failed with code '%d'",
				           status.si_status);
			else
				ok_printf("Child exited");
			break;

		case CLD_KILLED:
			err_printf("Child was terminated");
			break;

		default:
			err_printf("Child failed");
			break;
	}

	undo_cgroup(cgroup, ppid);

	return status.si_status;
}

static void do_daemonize(void) {
	int rc;

	openlog("pflask", LOG_NDELAY | LOG_PID, LOG_DAEMON);
	use_syslog = 1;

	rc = daemon(0, 0);
	if (rc < 0) sysf_printf("daemon()");
}

static void do_chroot(char *dest) {
	int rc;

	rc = chdir(dest);
	if (rc < 0) sysf_printf("chdir()");

	rc = chroot(".");
	if (rc < 0) sysf_printf("chroot()");

	rc = chdir("/");
	if (rc < 0) sysf_printf("chdir(/)");
}

static pid_t do_clone(void) {
	pid_t pid;

	pid = syscall(__NR_clone, clone_flags, NULL);
	if (pid < 0) {
		if (errno == EINVAL) {
			clone_flags &= ~(CLONE_NEWUSER);
			pid = syscall(__NR_clone, clone_flags, NULL);
		}
	}

	if (pid < 0) sysf_printf("clone()");

	return pid;
}

static inline void help(void) {
	#define CMD_HELP(CMDL, CMDS, MSG) printf("  %s, %-15s \t%s.\n", COLOR_YELLOW CMDS, CMDL COLOR_OFF, MSG);

	printf(COLOR_RED "Usage: " COLOR_OFF);
	printf(COLOR_GREEN "pflask " COLOR_OFF);
	puts("[OPTIONS] [COMMAND [ARGS...]]\n");

	puts(COLOR_RED " Options:" COLOR_OFF);

	CMD_HELP("--mount", "-m",
		"Create a new mount point inside the container");
	CMD_HELP("--netif", "-n",
		"Create a new network namespace and optionally move a network interface inside it");

	CMD_HELP("--user",  "-u",
		"Run the command as the specified user inside the container");

	CMD_HELP("--chroot",  "-r",
		"Use the specified directory as root inside the container");
	CMD_HELP("--chdir", "-c",
		"Change to the specified directory inside the container");

	CMD_HELP("--detach", "-d",
		"Detach from terminal");
	CMD_HELP("--attach", "-a",
		"Attach to the specified detached process");

	CMD_HELP("--setenv", "-s",
		"Set additional environment variables");

	puts("");

	CMD_HELP("--no-userns",  "-U", "Disable user namespace support");
	CMD_HELP("--no-mountns", "-M", "Disable mount namespace support");
	CMD_HELP("--no-netns",   "-N", "Disable net namespace support");
	CMD_HELP("--no-ipcns",   "-I", "Disable IPC namespace support");
	CMD_HELP("--no-utsns",   "-H", "Disable UTS namespace support");
	CMD_HELP("--no-pidns",   "-P", "Disable PID namespace support");

	puts("");
}
